controllers/phases.go

```diff
@@ -0,0 +1,477 @@
+	}
+}
+
+// newReconcilePhases returns phase reconciler for the given provider.
+func newReconcilePhases(r GenericProviderReconciler, provider genericprovider.GenericProvider, providerList genericprovider.GenericProviderList) *phaseReconciler {
+	return &phaseReconciler{
+		ctrlClient:           r.Client,
+		ctrlConfig:           r.Config,
+		certManagerInstaller: r.CertManagerInstaller,
+		clusterctlProvider:   &clusterctlv1.Provider{},
+		provider:             provider,
+		providerList:         providerList,
+	}
+}
+
+// preflightChecks a wrapper around the preflight checks.
+func (p *phaseReconciler) preflightChecks(ctx context.Context) (reconcile.Result, error) {
+	return preflightChecks(ctx, p.ctrlClient, p.provider, p.providerList)
+}
+
+// load provider specific configuration into phaseReconciler object.
+func (p *phaseReconciler) load(ctx context.Context) (reconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.V(2).Info("Loading provider", "name", p.provider.GetName())
+
+	// Load provider's secret and config url.
+	reader, err := p.secretReader(ctx, p.provider)
+	if err != nil {
+		return reconcile.Result{}, ifErrorWrapPhaseError(err, "failed to load the secret reader", operatorv1.PreflightCheckCondition)
+	}
+
+	// Initialize the a client for interacting with the clusterctl configuration.
+	p.configClient, err = configclient.New("", configclient.InjectReader(reader))
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Get returns the configuration for the provider with a given name/type.
+	// This is done using clusterctl internal API types.
+	p.providerConfig, err = p.configClient.Providers().Get(p.provider.GetName(), util.ClusterctlProviderType(p.provider))
+	if err != nil {
+		return reconcile.Result{}, ifErrorWrapPhaseError(err, operatorv1.UnknownProviderReason, operatorv1.PreflightCheckCondition)
+	}
+
+	spec := p.provider.GetSpec()
+
+	// If a configmap selector was specified, use it to find the configmap with provider configuraion. This is
+	// a case for "air-gapped" environments. If no selector was specified, use GitHub repository.
+	if spec.FetchConfig != nil && spec.FetchConfig.Selector != nil {
+		log.V(2).Info("Custom ConfigMap was provided for fetching manifests", "provider", p.provider.GetName())
+		p.repo, err = p.configmapRepository(ctx, p.provider)
+	} else {
+		p.repo, err = repository.NewGitHubRepository(p.providerConfig, p.configClient.Variables())
+	}
+	if err != nil {
+		return reconcile.Result{}, ifErrorWrapPhaseError(err, "failed to load the repository", operatorv1.PreflightCheckCondition)
+	}
+
+	// Store some provider specific inputs for passing it to clusterctl library
+	p.options = repository.ComponentsOptions{
+		TargetNamespace:     p.provider.GetNamespace(),
+		SkipTemplateProcess: false,
+		Version:             p.repo.DefaultVersion(),
+	}
+	if spec.Version != nil {
+		p.options.Version = *spec.Version
+	}
+
+	err = p.validateRepoCAPIVersion(p.provider)
```

Comment:
Stylistic, but in general it's much nicer to have an `if err != nil`, rather than just returning as we have here. It's cleaner to have a happy path `return reconcile.Result{}, nil` here IMO